### PR TITLE
User/seolive/jank events

### PR DESCRIPTION
**Pull request by seolive-msft**
PR for #41 adding Jank Frame table 
Adds PerfettoFrameEvent which shows scheduled and actually displayed frames in a single "Jank Event" table.

### CODE REVIEW

File: PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+{
+    /// <summary>
+    /// An event that represents a frame that was scheduled or actual displayed by a process. 
+    /// </summary>
+    public class PerfettoFrameEvent
+    {
+
+        public string FrameType { get; }
+        public string ProcessName { get; }
+        public long Upid { get; }
```
Comment (ivberg) [inline PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs]: Looks to be uint32 from docs - https://perfetto.dev/docs/analysis/sql-tables#expected_frame_timeline_slice

File: PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        public long DisplayToken { get; }
```
Comment (ivberg) [inline PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs]: Prefer we match name from doc aside from casing. E.g. DisplayFrameToken

File: PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace PerfettoCds.Pipeline.DataOutput
+{
+    /// <summary>
+    /// An event that represents a frame that was scheduled or actual displayed by a process. 
+    /// </summary>
+    public class PerfettoFrameEvent
+    {
+
+        public string FrameType { get; }
```
Comment (ivberg) [inline PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs:15]: Is this needed or we can sub-class this with an ExpectedFrameTimeLineSlice and ActualFrameTimeLineSlice?

File: PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        public string ProcessName { get; }
+        public long Upid { get; }
+        public long DisplayToken { get; }
+        public long SurfaceToken { get; }
+        public TimestampDelta Duration { get; }
+        public Timestamp StartTimestamp { get; }
+        public Timestamp EndTimestamp { get; }
+        public string JankType { get; }
+        public string JankTag { get; }
+        public string AppOnTime { get; }
```
Comment (ivberg) [inline PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs]: Missed this in doc?

File: PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            long displayToken,
+            long surfaceToken,
+            TimestampDelta duration,
+            Timestamp startTimestamp,
+            Timestamp endTimestamp,
+            string JankType,
+            string JankTag,
+            string AppOnTime,
+            string PresentType,
+            string GpuComposition,
```
Comment (ivberg) [inline PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs]: doc says this is int32?

File: PerfettoCds/Pipeline/Tables/PerfettoFrameTable.cs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            tableGenerator.AddColumn(SurfaceTokenColumn, baseProjection.Compose(x => x.SurfaceToken));
+            tableGenerator.AddColumn(PresentTypeColumn, baseProjection.Compose(x => x.PresentType));
+            tableGenerator.AddColumn(PredictionTypeColumn, baseProjection.Compose(x => x.PredictionType));
+            tableGenerator.AddColumn(GpuCompositionColumn, baseProjection.Compose(x => x.GpuComposition));
+            tableGenerator.AddColumn(JankTagColumn, baseProjection.Compose(x => x.JankTag));
+            tableGenerator.AddColumn(StartTimestampColumn, startProjection);
+            tableGenerator.AddColumn(EndTimestampColumn, endProjection);
+
+
+            var jankFramesConfig = new TableConfiguration("Jank Frames")
```
Comment (ivberg) [inline PerfettoCds/Pipeline/Tables/PerfettoFrameTable.cs]: We should play with any additional views might be useful as well

File: PerfettoProcessor/Events/PerfettoActualFrameEvent.cs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+                            Type = strVal;
+                            break;
+                        case "category":
+                            Category = strVal;
+                            break;
+                        case "layer_name":
+                            LayerName = strVal;
+                            break;
+                        case "jank_type":
+                            JankType = strVal;
```
Comment (ivberg) [inline PerfettoProcessor/Events/PerfettoActualFrameEvent.cs]: should intern the value

File: PerfettoProcessor/Events/PerfettoActualFrameEvent.cs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+                            break;
+                        case "jank_tag":
+                            JankTag = strVal;
```
Comment (ivberg) [inline PerfettoProcessor/Events/PerfettoActualFrameEvent.cs]: ditto - string intern the value

File: PerfettoProcessor/Events/PerfettoActualFrameEvent.cs [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+                            break;
+                        case "prediction_type":
+                            PredictionType = strVal;
```
Comment (ivberg) [inline PerfettoProcessor/Events/PerfettoActualFrameEvent.cs]: ditto

**Review by ivberg — CHANGES_REQUESTED** [review]

File: PerfettoUnitTest/PerfettoUnitTest.cs [hunk 10]
```diff
@@ -228,6 +231,28 @@ public void TestChromeTrace()
             Assert.IsTrue(logcatEventData.Count == 43);
             Assert.IsTrue(logcatEventData[0].Message == "type: 97 score: 0.8\n");
             Assert.IsTrue(logcatEventData[1].ProcessName == "Browser");
+        }
+
+        [TestMethod]
+        public void TestJankFrameTrace()
+        {
+            LoadTrace(@"..\..\..\..\TestData\Perfetto\jankFrame.pftrace");
```
Comment (ivberg) [inline PerfettoUnitTest/PerfettoUnitTest.cs:239]: This test file is 26MB in size and probably not appropriate in size to check into source code. Suggest you record a much smaller trace that is only a few MB in size.

File: PerfettoCds/Pipeline/Tables/PerfettoFrameTable.cs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+            {
+                Columns = new[]
+                {
+                    ProcessNameColumn,
+                    SurfaceTokenColumn,
+                    FrameTypeColumn,
```
Comment (ivberg) [inline PerfettoCds/Pipeline/Tables/PerfettoFrameTable.cs:123]: Going to leave some comments about changing the default view. Both to make it easier to use & to closer match the Perfetto UI. Changes can be made in the column but will put comments here for readability. FrameType should be sort descending 0.

File: PerfettoCds/Pipeline/Tables/PerfettoFrameTable.cs [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+                    TableConfiguration.PivotColumn, // Columns before this get pivotted on
+                    DisplayTokenColumn,
+                    PresentTypeColumn,
+                    AppOnTimeColumn,
+                    PredictionTypeColumn,
+                    JankTypeColumn,
+                    JankTagColumn,
+                    GpuCompositionColumn,
+                    ProcessIdColumn,
+                    DurationColumn,
```
Comment (ivberg) [inline PerfettoCds/Pipeline/Tables/PerfettoFrameTable.cs:132]: Should be sort descending order 1. Aggregation Max

**Review by ivberg — CHANGES_REQUESTED** [review]

File: PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs (hunk 4 above)
Comment (seolive-msft) [reply PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs]: Renamed to match docs

File: PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs (hunk 3 above)
Comment (seolive-msft) [reply PerfettoCds/Pipeline/DataOutput/PerfettoFrameEvent.cs:15]: Per our discussion, leaving as one class for code simplicity.

**Review by ivberg — APPROVED** [review]

**Pull request merged**